HTMLSerializer.js

```diff
@@ -117,12 +137,49 @@ var HTMLSerializer = class {
    */ 
   processDocument(doc) {
     this.html.push('<!DOCTYPE html>\n');
+    this.html.push(''); // Entry where pseudo element style tag will go.
     var nodes = doc.childNodes;
     for (var i = 0, node; node = nodes[i]; i++) {
       if (node.nodeType != Node.DOCUMENT_TYPE_NODE) {
         this.processTree(node, 0);
       }
     }
+    var pseudoElements = `<style>${this.pseudoElementCSS.join('')}</style>`;
+    this.html[1] = pseudoElements;
+  }
+
+  /**
+   * Takes an HTML element, and if it has pseudo elements listed in
+   * |this.PSEUDO_ELEMENTS| they will be added to |this.pseudoElementCSS|.
+   * Additionally, if |element| doesn't have an id it will be given one in
+   * |this.html|.
+   *
+   * @param {Element} element The Element whose pseudo elements will be
+   *     processed.
+   * @private
+   */
+  processPseudoElements(element) {
+    var win = element.ownerDocument.defaultView;
+    for (var i = 0, pseudo; pseudo = this.PSEUDO_ELEMENTS[i]; i++) {
+      var style = win.getComputedStyle(element, pseudo);
+      if (style.content) {
+        var nestingDepth = this.windowDepth(win);
+        var escapedQuote = this.escapedCharacter('"', nestingDepth);
+        var styleText = style.cssText.replace(/"/g, escapedQuote);
+        var id;
+        if (!element.attributes.id) {
+          do {
```

Comment:
Done - and yeah, this seems cleaner because generateId now actually returns a valid id.